packages/flame/lib/src/collisions/quadtree/has_quadtree_collision_detection.dart

```diff
@@ -0,0 +1,100 @@
+///
+mixin HasQuadTreeCollisionDetection on FlameGame
+    implements HasCollisionDetection<QuadTreeBroadphase<ShapeHitbox>> {
+  late QuadTreeCollisionDetection _collisionDetection;
+
+  @override
+  QuadTreeCollisionDetection get collisionDetection => _collisionDetection;
+
+  @override
+  set collisionDetection(
+    CollisionDetection<ShapeHitbox, QuadTreeBroadphase<ShapeHitbox>> cd,
+  ) {
+    if (cd is! QuadTreeCollisionDetection) {
+      throw 'Must be QuadTreeCollisionDetection!';
+    }
+    _collisionDetection = cd;
+  }
+
+  /// Initialise the QuadTree.
+  ///
+  /// - [mapDimensions] describes the collision area coordinates and size.
+  ///   Should match to game map position and size.
+  /// - [maxObjects] (optional) - maximum objects count in one quadrant.
+  /// - [maxLevels] (optional) - maximum nested quadrants.
+  /// - [minimumDistance] (optional) - specify minimum distance between objects
+  ///   to consider them as possibly colliding. You can also implement the
+  ///   [minimumDistanceCheck] if you need some custom behavior.
+  ///
+  /// The [onComponentTypeCheck] checks if objects of different types should
+  /// collide.
+  /// The result of the calculation is cached so you should not check any
+  /// dynamical parameters here, the function is intended to be used as pure
+  /// type checker.
+  /// It usually should not be overridden, see
+  /// [CollisionCallbacks.onComponentTypeCheck] instead
+  void initializeCollisionDetection({
+    required Rect mapDimensions,
+    double? minimumDistance,
+    int maxObjects = 25,
+    int maxLevels = 10,
+  }) {
+    _collisionDetection = QuadTreeCollisionDetection(
+      mapDimensions: mapDimensions,
+      maxDepth: maxLevels,
+      maxObjects: maxObjects,
+      onComponentTypeCheck: onComponentTypeCheck,
+      minimumDistanceCheck: minimumDistanceCheck,
+    );
+    this.minimumDistance = minimumDistance;
+  }
+
+  double? minimumDistance;
+
+  bool minimumDistanceCheck(Vector2 activeItemCenter, Vector2 potentialCenter) {
+    return minimumDistance == null ||
+        !((activeItemCenter.x - potentialCenter.x).abs() > minimumDistance! ||
+            (activeItemCenter.y - potentialCenter.y).abs() > minimumDistance!);
+  }
+
+  bool onComponentTypeCheck(PositionComponent one, PositionComponent another) {
+    var checkParent = false;
+    if (one is CollisionCallbacks) {
+      if (!(one as CollisionCallbacks).onComponentTypeCheck(another)) {
+        return false;
+      }
+    } else {
+      checkParent = true;
+    }
+
+    if (another is CollisionCallbacks) {
+      if (!(another as CollisionCallbacks).onComponentTypeCheck(one)) {
+        return false;
+      }
+    } else {
+      checkParent = true;
+    }
+    final oneParent = one.parent;
+    final anotherParent = another.parent;
+
+    if (checkParent &&
+        oneParent is PositionComponent &&
+        anotherParent is PositionComponent) {
+      return onComponentTypeCheck(oneParent, anotherParent);
```

Comment:
This shouldn't be needed if `hitboxParent` is used.

Reply:
Fixed, but it did not change things a lot, also there is now additional warning because of "protected" keyword on `hitboxParent`

Reply:
Ah, fixed that in #1928 
It does change more than what is visible, because now `CompositeHitbox`es will work in this solution too.

Reply:
#1928 is now merged, so if you update this branch against `main` the warning should go away.